### DISCUSSION TITLE
Derivatives at interfaces

### DISCUSSION THREAD

**Discussion by HelgeGehring** [kinnala/scikit-fem#895] [Q&A] [answered]
Hey,
at first, thanks for the great package!

I'm trying to calculate the modes of nanophotonic waveguides using scikit-fem. You can find my current implementation at https://github.com/HelgeGehring/waveguidemodes

For this I implemented the equations for the eigenvalue-problem:
![image](https://user-images.githubusercontent.com/42973196/169417277-db5ab890-e13b-4a5c-91d6-a69ca4644a98.png)

But here I get a problem:
-  If I use ElementTriP0() for the values of epsilon, all derivatives are zero, but the result keeps quite stable if I change the mesh
-  If I use ElementTriP1(), the derivatives are taken into account and the result looks like what I expect. But in this case the values of epsilon also smooth out over the edges of the waveguide and the results are not exact. But I think this is expected as I use the linear element.

Could you give me a hint how to implement the equation right? I'd expect I need to separate it into subdomains and add a term for the interfaces instead of including it in "operator". For epsilon I think I can stay with ElementTriP0, as its constant within the subdomain, right?

I also would think that the selection of ElementTriP1() can make problems at the interfaces, as the modes aren't continuous there. But if I separate the problem into subdomains it should also allow discontinuities there, right?

Would be awesome if you could help me how to implement this right :)

**Comment by kinnala** [comment, marked as answer]
I thought about this for a while. I think ElementTriP0 is definitely the way to go. I have some trouble interpreting the problem though. Since your epsilon is piecewise constant, the derivative is zero everywhere. However, you probably want to consider some sort of distributional derivative and then the derivative would be Dirac delta etc. I'm guessing what happens is that you'd get some additional term with the size of the jump. But I cannot wrap my head around it right now. Do you have any references for this equation which would have, e.g., alternative formulations?

**Reply by HelgeGehring** [reply]
Thanks for your answer!

The best derivation of the equation I found up to now is here: https://web.archive.org/web/20180127224243/https://www.home.uni-osnabrueck.de/phertel/aps/Dielectric_Waveguides.pdf on page 16 (3.6). There's also a second equivalent formulation (3.8) but I don't think that one would be easier to implement, as it still has derivatives of epsilon (?)
Like in the example on page 16 (and also in the 1D-examples in chapter 2.4) epsilon is stepwise. It's the squared refractive index of the material and as the waveguide has a certain shape the epsilon also changes step-wise.

The more general problem is described in http://ab-initio.mit.edu/book/photonic-crystals-book.pdf where page 9 (7) together with page 35 (11) describes the complete system and allows even solving modes for waveguide whose shape change in propagation direction. Using that would allow to get rid of the epsilon-derivatives, right?
But this formulation would require periodic boundary conditions. Would that be possible with self-defined meshes? I found only example 42 about that.
Furthermore, H would also change rapidly at the interface, how would it be possible to allow that?

**Reply by kinnala** [reply]
(7) looks like H(curl)-conforming eigenvalue problem. We have the lowest order Nedelec element for tetrahedral meshes to discretize fields like H in (7). However, perhaps you are not willing to solve the full 3D problem. I don't know anything about waveguides, but I did some random Googling and found https://fenicsproject.discourse.group/t/rectangular-waveguide-problem-chapter-34-fenics-book/2361 which I think solves similar 2D problem with curl's. Perhaps what you want to solve can also be written in similar form, (curl u, curl v) = (u, v)?

Some other references:
- https://bitbucket.org/fenics-project/dolfin/src/master/python/demo/undocumented/waveguide/demo_waveguide.py
- section 34: https://launchpadlibrarian.net/83776282/fenics-book-2011-10-27-final.pdf

**Reply by HelgeGehring** [reply]
Thanks for the answer! Seems more complicated than I thought :/ But I can also try to solve it like descried in the book!
Just adding these "Jump-terms" at the interfaces is not a solution? (As it would keep the implementation easier to understand I think)
Another formulation for the problem would be (as the divergence of E is 0) if that's easier to implement:
![image](https://user-images.githubusercontent.com/42973196/176567496-e3478536-fc76-4f11-8279-cf28a7d3476b.png)
(where n^2=epsilon)

Also I'm thinking about the 1D-case, which might be way easier for starting:
For the TE-polarization of the mode the equation looks like 
![image](https://user-images.githubusercontent.com/42973196/176558930-9b95edc7-171a-4280-94bb-3622edf522c2.png)
where epsilon can has steps again and epsilon_eff is the eigenvalue.
which doesn't require these jump terms and therefore seems quite easy to implement. If the field is TM polarized the equation changes to 
![image](https://user-images.githubusercontent.com/42973196/176559040-7212eeb7-4029-43cd-aad6-ae6ef19bf776.png)
where the epsilon has this jump again (which also leads to a discontinuity in the calculated H-field.
Is there a way to implement these equations directly or is it necessary to use this alternative formulation with the curls?

Thanks!

**Reply by HelgeGehring** [reply]
I've read deeper into the links you found and example "34.3.3 Shielded microstrip" in https://launchpadlibrarian.net/83776282/fenics-book-2011-10-27-final.pdf is quite close to what I was searching for, thanks again!
It seems, that I'd need a "Nedelec 1st kind H(curl)"-element, i.e. 2D version of ElementTetN0, right? (This one: https://defelement.com/elements/examples/triangle-Nedelec-1.html ?)
As discussed in #879 it seems to be only available for the 3D-case? For adding it, 2D-support would need to be added to ElementHcurl and a ElementTriN0 would need to be implemented, right?

**Reply by HelgeGehring** [reply]
@kinnala Thanks again for the nice library! I was wondering how complicated it would be to add these elements?

We're currently using skfem for our integrated photonic design library gdsfactory (see https://github.com/gdsfactory/gdsfactory/blob/master/gdsfactory/simulation/thermal/heater.py).
Currently we have already a thermal simulation (I'm not completely sure if the heating up is implemented right, would be super happy about any feedback :) )

We would like to implement the thermal solver together with the mode solver to have a multiphysics-simulation for nanophotonic phase shifters. We're also thinking about other solvers, e.g. electrostatics, semiconductor, process simulation.... scikit-fem is our preferred option because of its modularity, especially as it allows to leverage other libraries for solving/optimizing and we also see a path to run it on our TPUs.

Given this, we also plan on writing a paper on gdsfactory, alongside quite some people in industry/academia who are contributors ( see https://github.com/gdsfactory/gdsfactory). We'd be happy about help on the simulation side, the implementation and writing about it. Might be a great application for scikit-fem if you're interested?

**Reply by kinnala** [reply]
I can try adding this element, it should not be a huge task.

**Reply by kinnala** [reply]
Interesting project you have! I'll be sure to check it out. Just now I'm a bit busy with other writing projects but I can always try to help with finite elements.

**Reply by kinnala** [reply]
There is now WIP PR in #937 with absolutely no tests. If you want, you can participate to testing. I have some standard tests under `tests` directory that I will extend to this eleent. We also need some kind of simple example or manufactured solution.

**Reply by HelgeGehring** [reply]
Thanks for the quick help!
I've found a complete fenics implementation of the example in the book in http://diyhpl.us/~nmz787/pdf/Using_the_FEniCS_Package_for_FEM_Solutions_in_Electromagnetics.pdf (4.1) and I'm working on porting it to scikit-fem in https://github.com/kinnala/scikit-fem/pull/942 but I think I have not set the boundry conditions correctly and also I'm unsure how to configure the eigenvalue solver.
Could you have a look on that simple example? Thanks! :)

**Reply by HelgeGehring** [reply]
Thanks for the cleanup of the example!
At the moment I still see some discrepancies from the results from the paper. For the lowest non-zero eigenvalues the paper found values of 
![image](https://user-images.githubusercontent.com/42973196/197068685-6c97acd6-c356-4e3a-8a72-f6905c7ea68e.png)
I had a look on the complete spectrum of eigenvalues using
```
A, B, x, I = condense(A, B, D=basis.get_dofs())
lam, *x = eig(A.todense(), B.todense())
print(np.sort(lam))
```
and for the lowest non-zero values I get: 1.16864592,  4.54938381,  4.62492003,  5.92211760 (if I choose the discretisation like in the paper)
Would you have an idea what is going wrong here? (the values also change when I change the discretization)

![image](https://user-images.githubusercontent.com/42973196/197069007-ee3888be-6b14-4dcb-a37b-287559bd9677.png)
Might the difference be that they use higher order elements?
I.e. should we use ElementTriP3 and the second order Nedelec element? This one https://defelement.com/elements/examples/triangle-Nedelec-2.html ?

**Reply by kinnala** [reply]
I did some reading last night and found out the covariant Piola mapping is different in 3D and 2D. So I have to just implement ElementHcurl again for 2D.

**Reply by kinnala** [reply]
Now that I think about this problem, it seems that Et and Ez are completely independent. Why don't we solve them independently? Do they sometimes become coupled?

**Reply by kinnala** [reply]
Now I'm getting this:
```python
TE10 (calculated): 9.868943779049554
TE10 (analytical): 9.869604401089358
```
For some reason I cannot find any of the other modes.

**Reply by HelgeGehring** [reply]
Thank you so much! Like this it's a pretty nice example for the element!
I've implemented a version where both couple (they only decouple in the extreme case of being close to the cutoff) in https://github.com/HelgeGehring/waveguidemodes (the math is described in 34.1.2 in https://launchpadlibrarian.net/83776282/fenics-book-2011-10-27-final.pdf )
The mode looks exactly like it should look like, but I find a lot of modes with slightly varying eigenvalues. I'd guess that's because I took out the lagrangian multiplyers for enforcing div E = 0. I don't really know how to enforce it in the case where E_z is present. Could you help me here?
Besides that: Is there a way to plot a colormap of only the x/y component of E_t? Usually the modes have a almost negligible value in one direction, while the other component gives a good understanding of the mode.

**Reply by kinnala** [reply]
I'm not able to run the newest version of pygmsh because gmsh package does not work on my system. So I'm unable to create `mesh.msh` that is required to run your example.

But I think you should be just able to do this:
```python
basis = Basis(mesh, ElementTriN0() * ElementTriP1() * ElementTriP1())

@BilinearForm
def gauge(E_t, E_z, lam, v_t, v_z, mu, w):
    # set div E = 0 using a Lagrange multiplier
    return dot(grad(lam), v_t) + dot(E_t, grad(mu))
```
However, it is not clear to me that whether the Lagrange multiplier is needed in this coupled system. The idea for the Lagrange multiplier is that $E + \nabla \phi$ is another solution for any scalar potential $\phi$ so the scalar potential component of $E$ is forced to zero using the Lagrange multiplier.

**Reply by HelgeGehring** [reply]
Here the mesh.msh I'm using: [mesh.zip](https://github.com/kinnala/scikit-fem/files/9843600/mesh.zip)

I've just tried to add the lagrange multiplyer as you suggested, but it doesn't really help. Also the mode doesn't look any more as it should look like :/
Wouldn't the divergence need to also include E_z? it doesn't vary in direction of z, but due to 
![image](https://user-images.githubusercontent.com/42973196/197321730-a6ec2cd0-bcb0-4edc-b65b-6f08caa08c60.png)
it has a dependence on z.
From the physical interpretation div E = 0 makes sense, as there are no sources within the waveguide. And to me it seems like there's an additional unwanted degree of freedom in the system which leads to finding many similar modes.
Or is there another error which could lead to finding to many eigenvalues/eigenmodes?

![image](https://user-images.githubusercontent.com/42973196/197321851-6da56b91-2027-4c0c-bcd6-36dd048445cd.png)

Here a screenshot how it should look like in principal (the image is not perfect, as they do there some simplifications, it should be continuous in the area outside of the core, the y-component of E_t is plotted). In the given system there should be only one Eigegenmode with a N_eff greater than 1.444 (np.sqrt(lams) / k0 gives the N_eff)
Would you have another idea what is going wrong here?

**Reply by kinnala** [reply]
Thanks for the mesh and the reference. I'll try to see if I can figure out something.

**Reply by kinnala** [reply]
Here is a plotting snippet for the components:
```python
    plot_basis = Etbasis.with_element(ElementVector(ElementTriP1()))
    Etp = plot_basis.project(Etbasis.interpolate(Et))
    (Etpx, plotx_basis), (Etpy, ploty_basis) = plot_basis.split(Etp)

    plotx_basis.plot(Etpx, colorbar=True, shading='gouraud').show()
    ploty_basis.plot(Etpy, colorbar=True, shading='gouraud').show()
    Ezbasis.plot(Ez, colorbar=True, shading='gouraud').show()
```

I wasn't able to fix the multiple eigenvalues issue yet. It seems that setting the scalar potential component of $E_t$ to zero is not the right thing to do here and it is most likely due to the coupling. I don't know these type of problems really well so I cannot come up with a solution out of my head. Perhaps some literature related to $H(curl)$ discretizations will be able to answer the question.

**Reply by kinnala** [reply]
Btw, I noticed that you get almost the same result by not setting any boundary conditions at all. So it suggest that we might be constraining the wrong things.

**Reply by kinnala** [reply]
Now that I look at the formulation it already has these terms: $(\nabla E_z, v_t) + (E_t, \nabla v_z)$ so it seems to me that $E_z$ already serves as some kind of Lagrange multiplier.

**Reply by HelgeGehring** [reply]
Thanks for the snippet! Using that, the discontinuities at the left and right interface get a bit washed out. I've tried it with the P0-Element, then it looks right, but it doesn't get interpolated :/ Is there a way to use the P1 element, but disconnect it at the interfaces of the inner rectangle?

Might it be necessary to set $\text{div} (\vec{E})=0$ instead of setting $\text{div} (\vec{E_t})=0$ The latter one changes the equation I'd guess and the first one should only restrict the solutions to divergence-free ones (?) But here I'm unsure how to calculate that, as it depends on the eigenvalue :/ ($E_t$  as it is subsituted with $E_{t,\gamma}=\gamma E_t$ and the part regarding E_z as $E=(E_t+E_z)\text{e}^{-\gamma z}$ as written in 34.1.2 in https://launchpadlibrarian.net/83776282/fenics-book-2011-10-27-final.pdf)
Might it make sense to calculate the divergence of the eigenvectors to see if they fulfill $\text{div} \vec{E} = 0$ via $(\nabla E_z, v_t) + (E_t, \nabla v_z)$?
And which other things could we constrain here?

About the boundary conditions:
The given mesh/epsilons resemble a waveguide surrounded by silicon oxide. The light is expected to propagate within/around the high epsilon core, while the low epsilon surrounding is chosen large enough that the light wouldn't see the it's surface.
So, it fits to my expectations that not setting the boundary conditions only leads to negligible changes. Maybe comparing both might even allow to see if the surrounding silicon oxide was chosen large enough.
(The simplified cutoff-version was for a hollow rectangular RF-waveguide, where the surface matters)

**Reply by kinnala** [reply]
The discontinuous P1 element is `ElementDG(ElementTriP1())`. In this case I thin you'd just wrap it to `ElementVector(ElementDG(ElementTriP1()))`.

**Reply by HelgeGehring** [reply]
Using ElementDG gives somehow wrong images :/ But I'd guess high enough resulution and P0 will also work fine :)

I tested using the scipy.linalg.eig instead of the sparse solver and I get exactly the values I want. I've updated the script in https://github.com/HelgeGehring/waveguidemodes/blob/main/mode_solver.py
I'm using a lower resolution mesh for this: [mesh.zip](https://github.com/kinnala/scikit-fem/files/9864589/mesh.zip)

Besides the modes with positive eigenvalue, there are several eigenmodes with negative eigenvalue, which seem very localized and the ones I looked at seem to have E_t=0 and a non-zero E_z.
Might it be, that the solutions the sparse solver gives are just some superpositions of the correct modes with these localized modes?
Could we get rid of these modes using a lagrange multiplier which also considers E_z?

Would you have an idea how to get this running with the sparse solver?
Thanks :)

**Reply by HelgeGehring** [reply]
I've found a different formulation of the problem with the constraint for the divergence in https://doi.org/10.1080/02726340290084012 and implemented it in https://github.com/HelgeGehring/waveguidemodes/blob/main/mode_solver_2.py , but it shows the same behavior :/ Also there the eigenvalues calculated using the dense matrix look fine while the sparse solver only returns strange modes :/

**Reply by HelgeGehring** [reply]
I switched now to SLEPc with the configuration I found in https://fenicsproject.discourse.group/t/mixed-formulation-of-maxwells-equations/7814
The updated version you find here: https://github.com/HelgeGehring/waveguidemodes/blob/main/mode_solver_2.py

It now does exactly what I want, but I don't know why. Would you have an idea why both eigensolvers differ?

About the ElementDG: 
![image](https://user-images.githubusercontent.com/42973196/198174120-960012d3-0212-4852-b31a-e4ce9c47606d.png)
and the not interpolated version looks like
![image](https://user-images.githubusercontent.com/42973196/198174258-863cd73a-5d63-4319-bfd2-52a18580403e.png)
Would there be a solution to interpolate just within the subdomains?

**Reply by kinnala** [reply]
I'm not familiar enough with eigenvalue solvers to comment on why SLEPc gives better results.

**Reply by kinnala** [reply]
That plot looks strange indeed. I need to run it on my own laptop to see what is wrong. The code looks just fine to me. I'll report later if I can figure out something.

**Reply by HelgeGehring** [reply]
Thanks! Did you find an error in the ElementDG?

I'm combining the mode solving now with a thermal simulation of the environment which includes a structure to heat up a certain area. As the epsilon depends on the temperature, this changes the mode.
https://github.com/HelgeGehring/waveguidemodes/blob/main/waveguidemodes/thermal_transient.py

In principal it's like https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex19.py , but I'd need both cases, cooling down and heating up a by setting the power in a certain subdomain, like in https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex17.py
The simulation looks fine and the steady state looks the same, but there seems to be a factor between the two.
[mesh.zip](https://github.com/kinnala/scikit-fem/files/9925437/mesh.zip)
Could you give me here a tip how to include the joule heating in the Crank–Nicolson scheme?

Thanks!

**Reply by kinnala** [reply]
I'm unable to find mode_solver_2.py anymore so I cannot reproduce your result.

However, in principle `ElementDG(ElementTriP1())` visualization looks just fine when I try it using another example:
```python
In [1]: from skfem import *
In [2]: m = MeshTri().refined(4)
In [3]: basis = Basis(m, ElementDG(ElementTriP1()))
In [8]: basis.plot(basis.project(lambda x: x[0]*x[1] + (x[0] ** 2 + x[1] ** 2 <
   ...: 0.5)), shading='gouraud', cmap='viridis').show()
```
![Figure_1](https://user-images.githubusercontent.com/973268/199671397-fcd9277f-92d5-4db1-9f60-32437b596a27.png)

**Reply by kinnala** [reply]
It's not entirely clear to me what governing equations you are now solving with the time-dependent problem. And when you are saying that there is a "factor between the two", it is not clear what you are comparing against.

**Reply by HelgeGehring** [reply]
I've refactored the script quite a bit to make it easier usable - here you find the version I previously referenced: https://raw.githubusercontent.com/HelgeGehring/waveguidemodes/f185218abcc586c9ae94b70534f7667cbc399ebf/mode_solver_2.py
And here the mesh for this simulation:
[mesh (4).zip](https://github.com/kinnala/scikit-fem/files/9926815/mesh.4.zip)

**Reply by HelgeGehring** [reply]
I'm trying to solve
![image](https://user-images.githubusercontent.com/42973196/199676032-fa816952-4ba3-411f-8d17-b00a0f033bb9.png)
where I replace Q by the heat henerated by the joule heating. Similar to https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex17.py I only apply the joule heating on a subdomain.

The distribution of the heat looks nice, but the mean temperature seems wrong.
To cross-check, I start with the temperature distribution I calculate using the steady state equation and then I do the timestepping.
When I monitor the average temperature (first 20us the joule heating is like in the steady state, then 0, then starting at 100us like in the steady state again) I get for the mean temperature over time
![image](https://user-images.githubusercontent.com/42973196/199674872-86472c92-6632-41d5-94ba-4ae8dcf45ead.png)
I'd expect, that as I have the power in the beginning and in the end like in the steady state to have also the same average temperature, but here I see a drop in the mean temperature as soon as I do the timestepping.

**Reply by HelgeGehring** [reply]
If you'd like to run it without generating the mesh, you just need to comment the line `mesh_from_polygons(polygons, resolutions, filename='mesh.msh', default_resolution_max=.3)` out and use the mesh in [mesh.zip](https://github.com/kinnala/scikit-fem/files/9925437/mesh.zip).

**Reply by HelgeGehring** [reply]
The equation which is implemented is slightly simplified as I use the thermal_diffusivity which is defined as 
![image](https://user-images.githubusercontent.com/42973196/199681610-febfd2af-55eb-423b-a307-3a5af075c4d8.png)
to have less parameters.

**Reply by HelgeGehring** [reply]
I found a reference for the https://www-users.cse.umn.edu/~arnold/8445.f11/notes.pdf page 81 in the middle, so it seems fine to me how I added the term for the joule heating.
But like this the average temperature of the steady state depends on the thermal_diffusivity which shouldn't be the case, as it's there as a prefactor to both, the joule heating term and the diffusion term :/

**Reply by kinnala** [reply]
Did I understand correctly: you are worried because the algorithm you have implemented doesn't preserve the mean temperature?

I'd probably start by removing the penalty method (`penalize(...)`) for setting boundary conditions because that is an inconsistent technique and leads to all sorts of problems.

**Reply by kinnala** [reply]
Are you sure that before 20 us your steady state has converged? It looks to me that it's still changing.

**Reply by kinnala** [reply]
Perhaps you can just change `penalize` to `enforce` and everything works.

**Reply by HelgeGehring** [reply]
Exaktly, I'd expect, if I start with the steady state (which I calculate by finding the eigenvector with dT/dt=0) and I also keep the same current as in the calculation of the eigenstate, the derivative to be zero also in the case I do timestepping. Thus I'd expect the values not to change. Looking at the mean temperature is just to see a bit more the dynamics and if there's another steady state.

Using
```
basis.plot(temperatures[0] / temperatures[-1], colorbar=True)
```
![image](https://user-images.githubusercontent.com/42973196/200158351-d8a1f9af-632b-4807-929d-6638476f4253.png)
I've also looked how the steady state changes locally and it seems to be mostly a constant between the steady-state and the distribution it converges to (except at the bottom boundry, which I fix to zero, there it changes a bit more)


After the 20us the state didn't converge yet fully, but I'd like it to not change at all in the beginning, so seeing the drop is already a bad sign.
In the end of the simulation (200us) it seems to be converged and I'd guess that's also the state it would go to if I let it converge more in the beginning.

So the dynamic seems as it should be, just the factor between the steady state I get using the eigenmode calculation I do in https://github.com/HelgeGehring/waveguidemodes/blob/main/waveguidemodes/thermal.py and the steady state I find in the time stepping is what messes it up.
Also, as the thermal_diffusivity is in both, the diffusion term and also in the joule heating term, it should be a common prefactor and thus not change the state it's converging to. It should just change how fast it is converging. But when I play with the numbers I see also the temperature it converges to change. So there seems to be something wrong, too.

The timescale of needing a bit more than 20us to converge seems to be fine, I'm taking Fig. 1 in https://opg.optica.org/oe/fulltext.cfm?uri=oe-27-8-10456&id=408126 as an example for checking the simulation.


I've just tried changing penalize to enforce, but then the simulation diverges directly :/ Do I need to add something else to use enforce instead of penalize?

**Reply by kinnala** [reply]
I'm not sure I still understand what is being solved here. You didn't specify initial conditions or boundary conditions. Moreover, what is exactly "joule heating"?

**Reply by HelgeGehring** [reply]
Okay, sorry, in more detail:
The structure looks like this:

![image](https://user-images.githubusercontent.com/42973196/200159845-2ed36139-5fc8-40fc-bbc4-1e203f4c65c1.png)
The area named TiN is a metal, through which a current is run. As it has a resistance it heats up. Current^2 * resistance gives the power which is dissipated, this process is called joule heating (https://en.wikipedia.org/wiki/Joule_heating)

In the equation
![image](https://user-images.githubusercontent.com/42973196/200159912-3d4960c9-6f47-4f58-a36a-69d87d2395d1.png)
the Q is the applied heat energy transfer rate which is introduced in the system by this current.
Mathematically, I just have a "heating per area", which I need to apply to this area.

(The aim of this simulation is to get the heat profile in the Silicon (Si in the picture). The Silicon is used for guiding the light and I want to see how the heating is changing the local refractive index profile, and thus the eigenvalue of the Maxwell problem I needed the Nedelec element for - but this I'll add once the thermal simulation works fine)

Both, the TiN heater and the Si-Waveguide are surrounded by SiO2, which conducts the heat way worse than the TiN/Si.

Below the BOX in the image is a thick wafer of Silicon, i.e. a very good thermal conductor. Therefore, it's a good approximation to just fix the temperature at the interface from the BOX to the Silicon below to a certain temperature and don't simulate the Silicon below the BOX (As I'm only looking on offsets, I just fix it to 0)
Above the shown structure is Air, which conducts so bad, that it can be ignored.
To the left and right I just make the simulation so wide, that there is not any heat change at these sides, so I also don't need to care there about fixing the temperature.

**Reply by HelgeGehring** [reply]
One initial condition would be to set everything to zero, which would be the case if I just leave the system without any current for some time.

I'm here starting with the temperature profile I get using the steady state calculation, as it's also nice for testing if the time stepping is implemented correctly. If I keep the heating the same as in the calculation of the steady state I shouldn't see a change.

In the end I'm interested in simulating 
(1) the process where it heats up, i.e. initial condition everything 0 and see the time it needs to go to the steady state (with a current running through the heater, i.e. a local heating Q applied to the system).
(2) the process where it cools down, i.e. initial condition: steady state, see the time it needs to go to 0 (without any heating, i.e. Q=0).

**Reply by HelgeGehring** [reply]
The materials used are described mostly by two parameters:

The thermal conductivity, which defines the temperature profile in the steady state 
and
The thermal diffusivity, which is only plays a role in the time stepping simulation and defines how fast the steady state is reached.

I'm using a P0 element to set these values locally in each domain and then I interpolate them for assembling the matrices.

**Reply by HelgeGehring** [reply]
Using these two parameters the equations is
![image](https://user-images.githubusercontent.com/42973196/200162037-1a618847-b000-4864-ae86-9d358c8c3561.png)
where k is the thermal conductivity and alpha is the thermal diffusitivity

**Reply by kinnala** [reply]
Do I have it correct?
- Governing equation: $\frac{k}{\alpha} \frac{dT}{dt} - \nabla (k \nabla T) = Q$
- $T=0$ on the bottom boundary
- $k \nabla T \cdot n = 0$ on all other boundaries
- $k$ is elementwise constant, changes element-by-element
- $Q$ changes element-by-element and also in time to introduce those steps you showed earlier

**Reply by kinnala** [reply]
Why are you passing these keyword arguments to `mass`?
https://github.com/HelgeGehring/waveguidemodes/blob/main/waveguidemodes/thermal_transient.py#L45

The above definition of `mass` seems to imply that they are not used.

**Reply by HelgeGehring** [reply]
Your summary fits exactly to what I'm simulating - I think I should sum that all up in some documentation file.

One addition: $k$ and $\alpha$ are both within each subdomain constant, but change between the subdomains. I.e. "elementwise constant, changes element-by-element" also fits!

Your definition about Q is right, too! In the code Q is implemented as current_density**2/specific_conductivity, but that's just to have the physical parameter I can control (curreny_density) as the value which I change.

I've remved the two keywords from mass, that was just for some testing I did.

**Reply by kinnala** [reply]
Since those $k$ and $\alpha$ change in space, how exactly are you moving them out of the integral/derivative operators to have them in the load?

**Reply by kinnala** [reply]
I mean, they are not constant, so there is no rule which allows you to do this.

**Reply by HelgeGehring** [reply]
I just thought as they are element-wise constant I could just move them around like that.

But if it's not allow that would be great as it might then be the remaining mistake :)

How would I implement it correctly? Just leave both parameters in the equation where they should be?
How do I multiply the resulting term by $k/\alpha$?

**Reply by kinnala** [reply]
$k/\alpha$ is in the time derivative term so what if you just multiply the mass bilinear form by this? In the laplacian you'd have just k and the loading is just Q.

**Reply by HelgeGehring** [reply]
I'll try, but then it would not effect the part which is added in
```
M0 - (1 - theta) * L0 * dt
```
if I only modify M0, right?

**Reply by HelgeGehring** [reply]
Ah no, I see

**Reply by HelgeGehring** [reply]
You mean like this
```

    @BilinearForm
    def diffusivity_laplace(u, v, w):
        return w["thermal_conductivity"] * dot(u.grad, v.grad)

    @BilinearForm
    def mass(u, v, w):
        return w["thermal_conductivity"] / w["thermal_diffusivity"] * u * v
```
?

**Reply by kinnala** [reply]
Yes but I think ex17.py doesn't have the loading term so we'd need to implement that. Let me see how your implementation differs.

**Reply by kinnala** [reply]
So I was trying to derive the Crank-Nicholson step using this notation. (This example is not made by me so I'm a bit rusty.)

I got $(M + \frac{\delta}{2} L) x_k = (M - \frac{\delta}{2} L) x_{k-1} - \delta f$ which assumes $f$ stays constant, otherwise you need $f = \frac{1}{2}(f_k + f_{k-1})$.

**Reply by kinnala** [reply]
I started from the space discretized problem $Mx' + Lx + f = 0$.

**Reply by HelgeGehring** [reply]
Nice, that matches what is derived in https://www-users.cse.umn.edu/~arnold/8445.f11/notes.pdf on page 81
![image](https://user-images.githubusercontent.com/42973196/200251272-6fe9039b-d005-42fb-8181-130a01988164.png)
Implementing $f = \frac{1}{2}(f_k + f_{k-1})$ is no problem (I'll do that once it's working fine), but I'd guess without, it just shifts the time by $\delta/2$

**Reply by HelgeGehring** [reply]
I've updated https://github.com/HelgeGehring/waveguidemodes/blob/main/waveguidemodes/thermal_transient.py - it looks now nicer, but now it seems that it converges to a state whose average temperature is three times larger than the steady state :/ but it seems to be quite exact a factor of three.
Also if I locally compare the steady state and the state it converges to, it seems quite a constant factor (except close to the bottom boundary, but there the values are anyway close to 0, so the factor has probably a larger error)

**Reply by HelgeGehring** [reply]
Also now it works if I use `enforce` instead of `penalize`! But it produces the same result

**Reply by kinnala** [reply]
Perhaps it's clear to you, but penalize will modify the system by adding some large coefficients on the diagonal of L and M. I would try applying the BC T=0 directly on (A, B) the system somehow. You try to be energy conservative with Crank-Nicolson so penalize will ruin it for sure. This probably won't cause the 3x discrepancy though :-).

**Reply by HelgeGehring** [reply]
Found it! In the joule heating I forgot the `*v`.
It seems now to work perfectly! Thank you so much for your help! :)

**Reply by kinnala** [reply]
You are taking the transpose of A before solve. Is A really symmetric after enforce?

**Reply by kinnala** [reply]
Ok great!

**Reply by kinnala** [reply]
That must be one of the most common bugs. I really don't know how to detect that automatically. Perhaps parse the AST of every form and warn if v is not used at all.

**Reply by HelgeGehring** [reply]
Penalize is new to me, I mostly used it as it was used in https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex19.py and enforce gave me problems.
Where would I want to use penalize? Or is enforce always the better way to go? Would you have a literature reference?
I'm still learning how to use FEM and happy to understand more :)

**Reply by kinnala** [reply]
I wouldn't use it for other than quick tests. Mainly because the penalty parameter is quite random and too large values will deteoriate the condition number of the linear system. It ruins the numerical stability of many algorithms used for solving the linear system. However, it is slightly nontrivial to find exact conditions where this happens. It is better to not use it. Perhaps I'll change the example at some point.

**Reply by HelgeGehring** [reply]
Okay, then I'll stay away from `penalize` :)
What would be the reason to use it for quick tests instead of `enforce`?

**Reply by HelgeGehring** [reply]
And I've just checked, at least for the current settings L and M are symmetric
`print(np.sum(np.abs(L.T-L)), np.sum(np.abs(M.T-M)))` returned zeros
